### ISSUE TITLE
Installation issue: qt ^pkg-config when assimp is in environment with view

### ISSUE THREAD

**Issue by wdconinc** [LLNL/spack#28181]
### Steps to reproduce the issue

```console
cat << EOF > spack.yaml
spack:
  specs:
  - qt +opengl ^mesa -llvm
  - pkg-config
  - assimp
  concretization: together
  view: /opt/qt
EOF
spack env activate .
spack install
```
In this environment, `qt` fails to compile because it tries to link against the system `assimp`, which is not a formal dependency of `qt` but is found because `qt` queries `pkg-config` for assimp when compiling QtQuick3D when `+opengl`. The irony is that it works fine without `assimp` installed trough spack since then a 3rdparty version bundled with Qt is used (but in this case our environments need assimp for other reasons, i.e. `dd4hep`).

(Note: the `mesa -llvm` is optional here but makes this a more minimal example, timewise. OpenGL is required, though.)

Compilation error (full logs with `-j1`: [spack-build-01-configure-out.txt](https://github.com/spack/spack/files/7790000/spack-build-01-configure-out.txt), [spack-build-02-build-out.txt](https://github.com/spack/spack/files/7789997/spack-build-02-build-out.txt), [spack-build-env.txt](https://github.com/spack/spack/files/7790009/spack-build-env.txt)):
```
     7161    Project ERROR: Library 'assimp' is not defined.
  >> 7162    make[4]: *** [Makefile:73: sub-assimp-make_first] Error 3
     7163    make[4]: Leaving directory '/home/wdconinc/.spack/stage/spack-stage-qt-5.15.2-3sec4peaqtr34nbn4nk3lvyd5qrq2smv/spack-src/qtquick3d/src/plugins/assetimporters'
  >> 7164    make[3]: *** [Makefile:47: sub-assetimporters-make_first] Error 2
     7165    make[3]: Leaving directory '/home/wdconinc/.spack/stage/spack-stage-qt-5.15.2-3sec4peaqtr34nbn4nk3lvyd5qrq2smv/spack-src/qtquick3d/src/plugins'
  >> 7166    make[2]: *** [Makefile:182: sub-plugins-make_first-ordered] Error 2
     7167    make[2]: Leaving directory '/home/wdconinc/.spack/stage/spack-stage-qt-5.15.2-3sec4peaqtr34nbn4nk3lvyd5qrq2smv/spack-src/qtquick3d/src'
  >> 7168    make[1]: *** [Makefile:50: sub-src-make_first] Error 2
     7169    make[1]: Leaving directory '/home/wdconinc/.spack/stage/spack-stage-qt-5.15.2-3sec4peaqtr34nbn4nk3lvyd5qrq2smv/spack-src/qtquick3d'
  >> 7170    make: *** [Makefile:364: module-qtquick3d-make_first] Error 2
```

### Information on your system

* **Spack:** 0.17.1-675-84e26017b3
* **Python:** 3.9.7
* **Platform:** linux-ubuntu21.10-skylake
* **Concretizer:** clingo

### Additional information

Maintainer tag: @sethrj 

Not sure what the right approach is here. I guess one approach would be `depends_on('assimp', when='+opengl')` and fix the issue so it actually always compiles with the system (spack) assimp. Another would be `conflicts('assimp', when='+opengl')` and always use the 3rdparty version shipped with Qt.

I can emphatize with the trouble it must be to support Qt as a spack package: the build system tries hard to be smart and picks up all kind of things from the host system instead... and recompiling takes a long time.

What if no `pkg-config` as provider for `pkgconf`?
```yaml
spack:
  specs:
  - qt +opengl ^mesa -llvm
  - assimp
  concretization: together
  view: false
```
This picks up the host system `pkg-config` which doesn't know about `assimp` and therefore works fine (something else in our environments has a hard requirement on `pkg-config`).

Which other packages does `pkg-config` pick up and use, even if not explicited depended upon?
* `zstd`
* `dbus`

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by wdconinc** [comment]
Maybe what's needed is something like this:
```python
    depends_on("assimp", when='@5:+opengl')
    def common_config_args(self):
        if '+opengl' in spec:
            use_spack_dep('assimp')
```
(with some suitable version range where assimp is applicable as an option).

**Comment by sethrj** [comment]
Ah, the curse of vendored dependencies. You're right, the correct approach is to explicitly state the `assimp` dependency with `depends_on('assimp', when='+opengl')` and ensure Qt doesn't build its own. We should also do the same for zstd. I don't understand why `dbus` is still being picked up: the configure line should include `-no-dbus` to disable it, or use spack's if `+dbus` is used.

**Comment by wdconinc** [comment]
The configure script picks up dbus through pkg-config even when `qt -dbus` and `configure -no-dbus` because it looks for host_dbus:
```
Trying source 0 (type pkgConfig) of library host_dbus ...
+ /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/pkgconf-1.8.0-fkwoajbd3f3s575mk3rgsczkllnxqywj/bin/pkg-config --exists --silence-errors dbus-1 '>=' 1.2
+ /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/pkgconf-1.8.0-fkwoajbd3f3s575mk3rgsczkllnxqywj/bin/pkg-config --modversion dbus-1 '>=' 1.2
> 1.12.8
+ /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/pkgconf-1.8.0-fkwoajbd3f3s575mk3rgsczkllnxqywj/bin/pkg-config --libs-only-L dbus-1 '>=' 1.2
> -L/opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/dbus-1.12.8-gb5bbjs46xlgsh33fxqxqxtn4mjsw3pf/lib 
+ /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/pkgconf-1.8.0-fkwoajbd3f3s575mk3rgsczkllnxqywj/bin/pkg-config --libs-only-l dbus-1 '>=' 1.2
> -ldbus-1 
+ /opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/pkgconf-1.8.0-fkwoajbd3f3s575mk3rgsczkllnxqywj/bin/pkg-config --cflags dbus-1 '>=' 1.2
> -I/opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/dbus-1.12.8-gb5bbjs46xlgsh33fxqxqxtn4mjsw3pf/include/dbus-1.0 -I/opt/software/linux-ubuntu21.10-skylake/gcc-11.2.0/dbus-1.12.8-gb5bbjs46xlgsh33fxqxqxtn4mjsw3pf/lib/dbus-1.0/include 
 => source accepted.
test config.qtbase.libraries.host_dbus succeeded
```
Probably needs `-no-host_dbus` or so, but I don't see that listed in `configure --help`.

**Comment by sethrj** [comment]
Ugh... okeydoke. I'm honestly not sure what dbus does or the history of the option in the qt recipe.

**Comment by wdconinc** [comment]
I'm with you there :-) All I need qt for is geant4 :-)

**Comment by wdconinc** [comment]
Since qtquick3d was added in 5.14 (or that's the oldest tag on https://github.com/qt/qtquick3d), I imagine the version range would be:
```python
    depends_on("assimp", when='@5.14:+opengl')
    def common_config_args(self):
        if '+opengl' in spec:
            if version >= Version('5.14'):
                use_spack_dep('assimp')
```
Seems consistent with
```python
        if '~opengl' in spec:
            if version >= Version('5.14'):
                config_args.extend(['-skip', 'qtquick3d'])
```

**Comment by sethrj** [comment]
That sounds great. If you'll put in a PR I'll review it.

**Comment by sethrj** [comment]
Closed by #28191
